### PR TITLE
Display tasks in a gantt diagram #125

### PR DESCRIPTION
**Pull request by KebabRonin**
Add a new macro to display a Gantt Diagram.

image

### CODE REVIEW

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  #if($propValue)
+    $document.set($propName, $propValue)
+  #end
+#end
+
+#if("$!request.outputSyntax" == 'plain')
+  #set($discard = $response.setContentType('application/json'))
+#end
+
+#set($taskDocument = $request.taskDocument)
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:50]: I'll add for this page as well a comment in the beginning to explain the purpose of this page (as you do with comments at the beginning of a Java class)

Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:50]: Please check https://dev.xwiki.org/xwiki/bin/view/Community/VelocityCodeStyle/ for this page as well

**Review by oanalavinia — CHANGES_REQUESTED** [review]
Partial review, as I still need to check the JS code

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 2]
```diff
@@ -0,0 +1,2059 @@
...
+        popup.classList.add('gantt-popup');
+        ganttDiv.appendChild(popup);
+      }
+      // Set up popup event listeners
+      let progressChangeObserver;
+      let _hoverTimer;
+      let is_resizing_progress = false;
+
+      function showPopup(barElem) {
+        if (is_resizing_progress) return;
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:265]: It's better to always use `{}`, even if it's just a line
For example, someone could come and write another line after `return`, without realizing it will not be included in the `if` statement

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 3]
```diff
@@ -0,0 +1,2059 @@
...
+        const viewport = ganttDiv.getBoundingClientRect();
+        function normalizeDims(elem) {
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:267]: Thsi should be declared outside of this method

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 4]
```diff
@@ -0,0 +1,2059 @@
...
+          return {
+            x:(elem.x - viewport.x),
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:269]: `x : (elem.x - viewport.x)`
Same for below

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 5]
```diff
@@ -0,0 +1,2059 @@
...
+        `;
+        popup.classList.add('gantt-popup');
+        ganttDiv.appendChild(popup);
+      }
+      // Set up popup event listeners
+      let progressChangeObserver;
+      let _hoverTimer;
+      let is_resizing_progress = false;
+
+      function showPopup(barElem) {
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:264]: Thsi method should be declared outside

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 6]
```diff
@@ -0,0 +1,2059 @@
...
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </use>
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>require(['jquery', frappe_js_path], function(jQuery){
+  // Add custom popup and the custom svg markup when rendering Gantt
+  if(Gantt.prototype['_render'] === undefined) {
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:245]: I'll refactor this section of code, since it looks al mixed up and it's hard to follow (methods declarations and code are together).
So I'll have the methods declarations first, then the peace of code that actually uses it (check maybe [this](https://github.com/xwikisas/application-collabora/blob/main/application-collabora-ui/src/main/resources/Collabora/Code/UI.xml#L146))
If there are many methods, you could create modules based on their usage (e.g. some are only for popups, some for styling). Check for example [this](https://github.com/xwikisas/application-diagram/blob/master/application-diagram-ui/src/main/resources/Diagram/DiagramSheet.xml#L379) and [this](https://github.com/xwikisas/application-diagram/blob/master/application-diagram-ui/src/main/resources/Diagram/DiagramSheet.xml#L530), they both define different methods that will later be used, but are separated by their purpose and it becomes much easier to follow and understand the [code](https://github.com/xwikisas/application-diagram/blob/master/application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml#L1172). It might just be hard to identify these groups. Note that these modules can (and sometimes should) be in the same JS object, no need to create multiple ones. But you can still use one module in the other JS objects, if needed

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 7]
```diff
@@ -0,0 +1,2059 @@
...
+          hidePopup();
+        },
+        mousedown: function(e) {
+          clearTimeout(_hoverTimer);
+          _hoverTimer = null;
+          hidePopup();
+        }
+      });
+    }
+    function addStyles(svgElem) {
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:333]: Could you add a comment to explain why are these styles needed?

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 8]
```diff
@@ -0,0 +1,2059 @@
...
+        const b = normalizeDims(bar.$bar.getBoundingClientRect());
+        const pop = normalizeDims(popup.getBoundingClientRect());
+        let y = b.y - pop.height - 5;
+        let x = b.x + (b.width - pop.width) / 2;
+        popup.style.top  = y + 'px';
+        popup.style.left = x + 'px';
+        let popupNumber = jQuery('#gantt-popup-progress-number', popup)[0]
+        popupNumber.innerHTML = bar.compute_progress();
+        popup.style.visibility = 'visible';
+        if(progressChangeObserver) {
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:285]: Space missing `if (`

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 9]
```diff
@@ -0,0 +1,2059 @@
...
+    }
+    Gantt.prototype['_render'] = Gantt.prototype['render'];
+    Gantt.prototype['render'] = function() {
+      const result = this._render();
+      addStyles(this.$svg);
+      setupProgressPopup(this);
+      return result;
+    };
+  }
+  // Fix hour text display
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:368]: I'll elaborate. What exactly are you fixing?
Also, don't forget to add a `.` at the end of comments

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 10]
```diff
@@ -0,0 +1,2059 @@
...
+  if(Gantt.prototype['_get_date_info'] === undefined) {
+    Gantt.prototype['_get_date_info'] = Gantt.prototype['get_date_info'];
+    Gantt.prototype['get_date_info'] = function(date, last_date, i) {
+      let result = this._get_date_info(date, last_date, i);
+      const base_pos = {
+          x: i * this.options.column_width,
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:374]: Wrong indentation: it should only be 2ch long

**Review by oanalavinia — CHANGES_REQUESTED** [review]
I added some comments for the JS part as well
The main part is about the refactoring that is needed, in order to group the code a little. You'll have to check all code to see exactly what would be best fitted. I'll check the code in more detail after
Another aspect it's about code style. For this as well I added only some comments, you'll have to apply the suggestion to all (indentation should be 2characters, you are missing some spaces between `if`, `for` and the `(`, you are missing a `;` for `function() {};`, comments should end with `.`, ...). Even though it might not look like it, code style it's important for development, since it makes the code more clearer for you and the other devs that might need to work on it as well. Probably you have seen on older projects how hard it can be when code style is not respected. Unfortunately, there isn't a wiki page with JS specific rules.. But [this](https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/XhtmlCssCodeStyle/) one is of help

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/WebHome.xml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>true</hidden>
+  <content>The **taskgantt** macro draws an interactive Gantt diagram which shows tasks, optionally filtered by start/end date, assignee or project.
+
+=== Examples ===
+
+(% class="box" %)
+(((
+~{~{taskgantt readonly="false"/}} - enable editing through the UI (disabled by default)
```
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/WebHome.xml:45]: This is part of the example usage section, so it prevents executing the macro and displays it as plain text.

I could switch to using the {{code}} macro. But if I do, I don't think there is a way to use translation keys for the parameter descriptions without using velocity.

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+      $jsontool.serialize({'success':false, 'reason':$services.localization.render('taskmanager.gantt.taskEdit.error.permissions')})
+    #else
+      #set ($discard = $taskDocument.use('TaskManager.TaskManagerClass'))
+      #if ($request.endDate)#set ($end = $datetool.toDate($numbertool.toNumber($request.endDate).longValue()))#end
+      #if ($request.startDate)#set ($start = $datetool.toDate($numbertool.toNumber($request.startDate).longValue()))#end
+
+      #maybeSet ($taskDocument 'progress' $request.progress)
+      ## Set status automatically when 100%
+      #if ($request.progress == '100')
+        $taskDocument.set('status', 'Done')
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:84]: The best practice is to use a `#set ($discard = ...`

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+  <version>1.1</version>
+  <title/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>true</hidden>
+  <content>{{velocity wiki="false"}}
+#macro (maybeSet $document $propName $propValue)
+  #if ($propValue)
+    $document.set($propName, $propValue)
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:42]: The best practice is to use a `#set ($discard = ...`

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+  <author>xwiki:XWiki.Admin</author>
+  <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
+  <version>1.1</version>
+  <title/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>true</hidden>
+  <content>{{velocity wiki="false"}}
+#macro (maybeSet $document $propName $propValue)
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:40]: You could move this macro in a `{{velocity output="false"}}{{/velocity}}` code block,

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+##
+## Parameters:
+##   - taskDocument: document id (eg. wiki:Space.Page);
+##   - startDate, endDate: new times, given as unix epoch milliseconds;
+##   - progress: new progress value, 0-100 number;
+##
+## Returns a json with a 'success' property and, in the case of a failure, a 'reason' as defined in the translation keys.
+##
+
+
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:59]: To many newlines

Aren't these displayed in the response?

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+#if ("$!request.outputSyntax" == 'plain')
+  #set ($discard = $response.setContentType('application/json'))
+#end
+
+#set ($taskDocument = $request.taskDocument)
+#if (!$xwiki.exists($taskDocument))
+  $jsontool.serialize({'success':false, 'reason':$services.localization.render('taskmanager.gantt.taskEdit.error.notExists')})
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:66]: Line too long

```
  $jsontool.serialize({
    'success':false,
    'reason':$services.localization.render('taskmanager.gantt.taskEdit.error.notExists')
  })
```

I see the same for the other ` $jsontool.serialize` calls below. It's also easier to read when there is a structure and not in just one line

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+-->
+
+<xwikidoc version="1.4" reference="TaskManager.GanttMacro.UpdateService" locale="">
+  <web>TaskManager.GanttMacro</web>
+  <name>UpdateService</name>
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:25]: WDYT about `TaskUpdateService`? I'm not sure if it's not too much

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+  <author>xwiki:XWiki.Admin</author>
+  <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
+  <version>1.1</version>
+  <title/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>true</hidden>
+  <content>{{velocity wiki="false"}}
+#macro (putOrDefault $map $propertyName $value $default1)
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:40]: I'll add macro definitions to a `{{velocity output="false"}}{{/velocity}}` code block above this, for better structuring the code

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+  #set ($tasks = [])
+  #foreach ($taskDocument in $queryResult)
+    #set ($taskXDoc = $xwiki.getDocument($taskDocument))
+    #set ($task = $taskXDoc.getObject('TaskManager.TaskManagerClass'))
+    #if ($services.security.authorization.hasAccess('view', $taskDocument))
+      #set ($taskObj = {})
+      #set ($customClasses = '')
+      #set ($discard = $taskObj.put('id', "$taskXDoc.getDocumentReference()"))
+      #set ($discard = $taskObj.put('link', $taskXDoc.getURL()))
+      #putOrDefault ($taskObj 'actualName'
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:152]: You cannot use `name` directly?

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+      #putOrDefault ($taskObj 'start'
+        $task.getValue('startDate').getTime()
+        $task.getValue('createDate').getTime()
+      )
+      #putOrDefault ($taskObj 'end'
+        $task.getValue('duedate').getTime()
+        $xwiki.jodatime.getDateTime($taskObj['start']).plusWeeks(1).getMillis()
+      )
+      #if ("$!task.getValue('duedate').getTime()" == '')
+        #set ($customClasses = $customClasses + ' taskgantt-no-due-date')
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:172]: Maybe you could have a comma separated list of custom classes. I think it will be more intuitive to split then in JS by `,` instead of space.
Here, in velocity, you will have a list `#set ($customClasses = [])` where it's easier to add values than concatenating values. At the end you can use `$stringtool.join($customClasses, ',')`

I think this will also remove the need for `$task.getValue('status').replace(' ', '-')` below, right?

File: application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml [hunk 21]
```diff
@@ -334,6 +335,12 @@
         var status = $(this).val();
         if (status == "Done") {
           $("input[name='TaskManager.TaskManagerClass_0_progress']").val(100);
+        } else if (status == "InProgress") {
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml:338]: Single quotes should be used for simple strings

File: application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml [hunk 22]
```diff
@@ -334,6 +335,12 @@
...
+          let startDateInput = $("input[name='TaskManager.TaskManagerClass_0_startDate']");
+          const format = moment().toMomentFormatString(startDateInput.data("format"));
+          if (startDateInput.val() == "") {
+            startDateInput.val(moment().format(format));
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml:342]: Here you are simply the current date as the start date when status was changed to in progress?

**Review by oanalavinia — CHANGES_REQUESTED** [review]
Added review and some questions. Looks good overall :)  There are still some structure changes that I think are needed, but nothing too complex

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 23]
```diff
@@ -0,0 +1,2511 @@
...
+  #if ($xcontext.macro.params.from)
+    data-date-from="$escapetool.html($xwiki.jodatime.dateTime.parse($xcontext.macro.params.from, $dateTimeFormatter).getMillis())"
+  #end
+  #if ($xcontext.macro.params.to)
+    data-date-to="$escapetool.html($xwiki.jodatime.dateTime.parse($xcontext.macro.params.to, $dateTimeFormatter).getMillis())"
+  #end
+  #if ($xcontext.macro.params.hideNoDueDate)
+    data-hide-no-due-date="$escapetool.html($xcontext.macro.params.hideNoDueDate)"
+  #end
+  #if ($xcontext.macro.params.projects)
```
[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:1870]: Actually, we should open a ticket for this once the PR is merged. So we don't forget about it :)

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml (hunk 19 above)
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:152]: ```name``` is a field also used by the Gantt library.

The only use for having 2 name parameters is for removing a warning from the displayed name when a task with no due date is modified (only happens in the ```formatTask``` function in js).

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml (hunk 20 above)
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:172]: It can be changed, but the ```$task.getValue('status').replace(' ', '-')``` would still need to be there, either in velocity or JS.

Why the replace is needed: The ```custom_class``` parameter in the Gantt library is directly inserted as the class attribute in the task bar. So if a status has a space, its associated class would be interpreted as 2 different classes (Ex: status "To Do" -> class "taskgantt-status-To Do", which means the element will have css classes "taskgantt-status-To" and "Do").

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 24]
```diff
@@ -0,0 +1,2511 @@
...
+  };
+
+  return {
+    'updateService': updateService,
+    'getJsonSourceQuery': getJsonSourceQuery,
+  };
+});
+
+define('taskgantt-utils', ['jquery', 'taskgantt-localization'], function ($, l10n) {
+  function escapeHTML(originalString) {
```
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:198]: I only found the deprecated ```escape()``` method in JS, the next closest thing would be ```encodeURI()```.

Using jQuery's ```text()``` (or plain JS), there is the workaround of using a dummy element to paste the text into in order to have it escaped.

Most solutions I found online are writing a variation of this function.

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 25]
```diff
@@ -0,0 +1,2511 @@
...
+    Object.assign(result, tweakedPos);
+    return result;
+  };
+  /**
+   * Custom implementation of bind_bar_events from the frappe-gantt(v0.6.1) library.
+   * Don't move dependencies when moving a task.
+   * This prevents unintentional updates and more freedom when moving tasks.
+   */
+  function custom_bind_bar_events() {
+    // Redefine some jQuery functions so they work on our SVG elements.
```
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:664]: The Gantt library redefines some jQuery functions, probably in order to remain 'dependency free'.

I'll have to test if jQuery has any edge cases which come into play here, like the jQuery [on()](https://api.jquery.com/on/) function has this:
```Note: Delegated event handlers do not work for SVG.```

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml (hunk 15 above)
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/UpdateService.xml:59]: They only seem to appear when viewing the page through the XWiki UI, accessing with GET like in the diagram has no new lines, just the JSON.

File: application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml (hunk 22 above)
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml:342]: The ```startDate``` field was added in this PR, so could it be considered as part of the Gantt update?

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 26]
```diff
@@ -0,0 +1,2511 @@
...
+    new XWiki.Document(XWiki.Model.resolve('TaskManager.GanttMacro.UpdateService', XWiki.EntityType.DOCUMENT)).getURL('get')
+    + '?xpage=plain&amp;outputSyntax=plain&amp;';
+
+  function updateService(task) {
+    // Could be refactored and replaced by a REST API call after XWIKI-20704 is fixed.
+    return updateServiceURL + new URLSearchParams({ taskDocument: task.id }) + '&amp;';
+    //return new XWiki.Document(XWiki.Model.resolve(task.id, XWiki.EntityType.DOCUMENT)).getRestURL('objects/TaskManager.TaskManagerClass/0');
+  };
+  function getJsonSourceQuery(ganttDiv) {
+    let query = '';
```
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:165]: $.params automatically does ```encodeURIComponent```, having both breaks when tasks have special characters.

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 27]
```diff
@@ -0,0 +1,2511 @@
...
+  #if ($xcontext.macro.params.projects)
+    data-projects="$escapetool.html($xcontext.macro.params.projects)"
+  #end
+  #if ($xcontext.macro.params.assignees)
+    data-assignees="$escapetool.html($xcontext.macro.params.assignees)")
+  #end
+  #if ($xcontext.macro.params.reporters)
+    data-reporters="$escapetool.html($xcontext.macro.params.reporters)")
+  #end
+  #if ($xcontext.macro.params.spaces)
```
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:1879]: Is it useful to have the parameter accept multiple spaces? I could change it to only have one space and use the existing picker.

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml (hunk 20 above)
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:172]: To ask Stefana is we need the custom classes for toDo, inProgress and Done

File: application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml (hunk 22 above)
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/TaskManagerSheet.xml:342]: Yes, it's ok to be part of this PR

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml (hunk 25 above)
[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:664]: Since this method is copied from frappe-gantt and has only some things changed, there should be comments added for the exact thing that were changed. This will make future upgrades easier

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml (hunk 24 above)
[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:198]: It's ok to keep this method

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+#macro (putOrDefault $map $propertyName $value $default1)
+  #if ($value)
+    #set ($discard = $map.put($propertyName, $value))
+  #else
+    #set ($discard = $map.put($propertyName, $default1))
+  #end
+#end
+
+#macro (csvToArray $csvString $return)
+  ## When a project name contains a comma, it can be escaped with a backslash.
```
[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml:49]: TODO for me: to investigate what is the best practice for fields with comma separated values

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml (hunk 27 above)
[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:1879]: Todo: ask Stefana if we should have `space` or `spaces`

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml (hunk 28 above)
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml]: Update: I [implemented](https://github.com/KebabRonin/application-task/tree/parameter-list-escapes-beta) the backslash escapes, but even though it works for now, backslashes inside names can still lead to some unforeseen consequences:
* REST APIs not working (tested by trying to update a task status in the `kanbanboard` macro)
* The `$xwiki.getURL()` method returns the formatted names, so links to spaces which conflict with the naming strategy lead to non-existent pages
  * Since the default behavior is to not check page titles on save, creating a new page through a non-existent url will succeed, even if the name contains (url escaped) backslashes

File: application-task-ui/src/main/resources/TaskManager/TaskManagerTranslations.xml [hunk 29]
```diff
@@ -117,6 +117,43 @@ taskmanager.livetable.assignee=Assignee
 taskmanager.livetable.duedate=Due Date
 taskmanager.livetable.creationDate=Creation Date
 
+# Gantt Diagram
+taskmanager.gantt.taskEdit.error.permissions=You do not have sufficient permissions to modify this task
+taskmanager.gantt.taskEdit.error.generic=The task could not be updated
+taskmanager.gantt.taskEdit.error.readonly=This diagram is read only
+taskmanager.gantt.taskEdit.error.notExists=The task does not exist
+taskmanager.gantt.taskEdit.error.notExistsClass=The task class does not exist
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/TaskManagerTranslations.xml:125]: Task class does not exist

**Review by oanalavinia — CHANGES_REQUESTED** [review]

**Comment by oanalavinia** [comment]
We had a meeting to discuss the latest changes / review (everything is added as comments above)

The other PRs were also discussed

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml (hunk 28 above)
[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/JSONTaskSource.xml]: Todo: check if `$wikimacro.parameters` can be used https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/Tutorials/WritingMacros/WikiMacroTutorial/ , or it is not working because it cannot deserialize the String value

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml [hunk 30]
```diff
@@ -0,0 +1,2541 @@
...
+  #set ($diagramIsReadonly = !("$!xcontext.macro.params.readonly" == 'false'))
+  ## Create HTML templates for use in JS.
+  #set ($config = {})
+  #set ($discard = $config.put('svgCustomDefs', "#svgCustomDefs"))
+  #set ($discard = $config.put('noTasksMessage', "#noTasksMessage"))
+  {{html clean='false'}}
+  &lt;script id='taskgantt-config-templates' type='application/json'&gt;$jsontool.serialize($config).replace('&amp;lt;', '\u003C')
+  &lt;/script&gt;
+  {{/html}}
+
```
[the remark this answers is not in the document]
Comment (KebabRonin) [reply application-task-ui/src/main/resources/TaskManager/GanttMacro/Macro.xml:1869]: No, it is not. While testing this I also found that the `{{/velocity}}{{velocity}}` line above breaks the missing licenses macro, so I'll add a new line between them in the next commit.

File: application-task-ui/src/main/resources/TaskManager/GanttMacro/TaskUpdateService.xml [hunk 31]
```diff
@@ -169,9 +163,6 @@
     <property>
       <allow>1</allow>
     </property>
-    <property>
-      <groups>XWiki.XWikiAllGroup</groups>
```
Comment (oanalavinia) [inline application-task-ui/src/main/resources/TaskManager/GanttMacro/TaskUpdateService.xml]: You still have the `XWiki.XWikiRights` objects, even if there are no groups specified. Are these needed?

**Review by oanalavinia — APPROVED** [review]
Looks good, can be merged! Nice work! :)

P.S. please don't forget to use **squash and merge** and to add an extended commit message
After merging it, the other issues should be opened (the ones that have TODOs inside the code)
